File: anagram-detector/typescript/src/anagramDetector.ts
function key(s: string): string {
  return s
    .toLowerCase()
    .split('')
    .filter((c) => c >= 'a' && c <= 'z')
    .sort()
    .join('');
}

export function areAnagrams(a: string, b: string): boolean {
  if (a === b) return false;
  const keyA = key(a);
  const keyB = key(b);
  if (keyA.length === 0 || keyB.length === 0) return false;
  return keyA === keyB;
}

export function findAnagrams(subject: string, candidates: readonly string[]): string[] {
  return candidates.filter((candidate) => areAnagrams(subject, candidate));
}

export function groupAnagrams(words: readonly string[]): string[][] {
  const keyToGroup = new Map<string, string[]>();
  const order: string[] = [];
  for (const word of words) {
    const k = key(word);
    let group = keyToGroup.get(k);
    if (group === undefined) {
      group = [];
      keyToGroup.set(k, group);
      order.push(k);
    }
    group.push(word);
  }
  return order.map((k) => keyToGroup.get(k)!);
}

File: anagram-detector/python/src/anagram_detector/anagram_detector.py
from typing import Iterable


def _key(s: str) -> str:
    return "".join(sorted(c for c in s.lower() if c.isalpha()))


def are_anagrams(a: str, b: str) -> bool:
    if a == b:
        return False
    key_a = _key(a)
    key_b = _key(b)
    if not key_a or not key_b:
        return False
    return key_a == key_b


def find_anagrams(subject: str, candidates: Iterable[str]) -> list[str]:
    return [candidate for candidate in candidates if are_anagrams(subject, candidate)]


def group_anagrams(words: Iterable[str]) -> list[list[str]]:
    key_to_group: dict[str, list[str]] = {}
    order: list[str] = []
    for word in words:
        k = _key(word)
        if k not in key_to_group:
            key_to_group[k] = []
            order.append(k)
        key_to_group[k].append(word)
    return [key_to_group[k] for k in order]
